openedx/core/djangoapps/enrollments/tests/test_views.py

```diff
@@ -229,6 +229,72 @@ def test_enroll(self, course_modes, enrollment_mode):
         assert is_active
         assert course_mode == enrollment_mode
 
+    def test_enroll_with_email_staff(self):
+        # Create enrollments with email are allowed if you are staff.
+
+        self.client.logout()
+        AdminFactory.create(username='global_staff', email='global_staff@example.com', password=self.PASSWORD)
+        self.client.login(username="global_staff", password=self.PASSWORD)
+
+        resp = self.client.post(
+            reverse('courseenrollments'),
+            {
+                'course_details': {
+                    'course_id': str(self.course.id)
+                },
+                'email': self.user.email
+            },
+            format='json'
+        )
+        assert resp.status_code == status.HTTP_200_OK
+
+    @patch('openedx.core.djangoapps.enrollments.views.EnrollmentListView.has_api_key_permissions')
+    def test_enroll_with_email_server(self, has_api_key_permissions_mock):
+        # Create enrollments with email are allowed if it is a server-to-server request.
+
+        has_api_key_permissions_mock.return_value = True
+        resp = self.client.post(
+            reverse('courseenrollments'),
+            {
+                'course_details': {
+                    'course_id': str(self.course.id)
+                },
+                'email': self.user.email
+            },
+            format='json'
+        )
+        assert resp.status_code == status.HTTP_200_OK
+
+    def test_enroll_with_email_without_staff(self):
+        # If you are not staff or server request you can't create enrollments with email.
```

Comment:
Could we add 3 tests similar to the ones above, but using parameters that test the following user stores?

* A non-staff/server request tries to...
  * enroll with `'username': self.user.username` → 200 OK
  * enroll without specifying username → 200 OK
  * enroll with `'username': self.user.username` but `'email': self.other_user.email` → 404 Not Found

Why: I suspect the latest commits may cause a regression where a user who tries to self-enroll without a username will be given a 404 when they should be given a 200. Adding these tests may be a quick way to rule this out.

Reply:
@pshiu, you were right; I had something that prevented a user from enrolling itself. I added the first two tests you mentioned and edited the last one because, in the case you said, the username has priority over the email, and the result is similar to case 1 you mentioned. Your last point is like the `test_enroll_with_user_and_email` test.

Reply:
Brilliant. Thanks for adding the tests.